include/ada/url.h

```diff
@@ -94,23 +94,43 @@ namespace ada {
     /**
      * A URL includes credentials if its username or password is not the empty string.
      */
-    bool includes_credentials() const {
+    [[nodiscard]] bool includes_credentials() const {
       return !username.empty() || !password.empty();
     }
 
     /**
      * A URL is special if its scheme is a special scheme. A URL is not special if its scheme is not a special scheme.
      */
-    bool is_special() const {
+    [[nodiscard]] bool is_special() const {
       return ada::scheme::is_special(scheme);
     }
 
     /**
      * A URL has an opaque path if its path is a string.
      */
-    bool has_opaque_path() const {
+    [[nodiscard]] bool has_opaque_path() const {
       return path.is_opaque();
     }
+
+    /**
+     * @see https://url.spec.whatwg.org/#shorten-a-urls-path
+     *
+     * This function assumes url does not have an opaque path.
+     */
+    void shorten_path() {
+      // Let path be url’s path.
+      // If url’s scheme is "file", path’s size is 1, and path[0] is a normalized Windows drive letter, then return.
+      if (scheme == "file" && path.list_value.size() == 1 && checkers::is_normalized_windows_drive_letter(path.list_value[0])) {
+        return;
+      }
+
+      // Remove path’s last item, if any.
+      path.list_value.pop_back();
+    }
+
+    [[nodiscard]] std::optional<uint16_t> scheme_default_port() const {
+      return scheme::SPECIAL_SCHEME.find(scheme)->second;
```

Comment:
Do you use some specific form of assertion? It might be good to verify that `.find()` actually returns a result.

Reply:
You're right. I missed this thinking `find()` returns `std::optional`, but value of SPECIAL_SCHEME is std::optional. I'm updating it now. Thanks!